wordpress-mobile/WordPress-Android
# [Modal Layout Picker] "Preview" button functionality

**Fixes:** https://github.com/wordpress-mobile/gutenberg-mobile/issues/2452

**Related PRs:**

- Depends on: https://github.com/wordpress-mobile/WordPress-Android/pull/13018

- `gutenberg`:
1. https://github.com/WordPress/gutenberg/pull/25661
1. https://github.com/WordPress/gutenberg/pull/25930
- `gutenberg-mobile`:
1. https://github.com/wordpress-mobile/gutenberg-mobile/pull/2674
1. https://github.com/wordpress-mobile/gutenberg-mobile/pull/2696
- `WordPress-FluxC-Android`
1. https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1715

To test:

Layout Picker should show [when creating a new page from *My Site* or *Site Pages* the *Modal Layout Picker* appears](https://github.com/wordpress-mobile/WordPress-Android/pull/12482).

- Start by navigating to the Modal Layout Picker

### Preview a layout
1. Select a layout from the list
1. Select Preview
- **Expect** to see a read-only editor load that shows a preview of the selected layout.

### Preview a layout, then a different one
1. Select a layout from the list
1. Select Preview
- **Expect** to see a read-only editor load that shows a preview of the selected layout.
1. Select Back
1. Select a different layout from the list
1. Select Preview
- **Expect** to see a read-only editor load that shows a preview of the newly selected layout.

### Preview a layout, then cancel
1. Select a layout from the list
1. Select Preview
- **Expect** to see a read-only editor load that shows a preview of the selected layout.
1. Select Back
- Or close the modal by swiping down from the top
1. Close the modal via the ✖️ 

### Preview a layout, then create a page
1. Select a layout from the list
1. Select Preview
- **Expect** to see a read-only editor load that shows a preview of the selected layout.
1. Select Create Page
- **Expect** to see the page editor load with the selected layout content populated.
- **Expect** to see the page editor load with a title that relates to the content. Example a layout from the "About Pages" should have a title like "About"

### Spot check for regressions
Open an existing page or blog, expect the content to be loaded and for normal editing to take place.

### Spot check Modal Layout Picker Disabled
1. Disable Modal Layout Picker
1. Create a new page
- **Expect** to see the existing Starter Page Template Picker
1. Select an option
- **Expect** to see the Preview with "Close" and "Apply" as actions
1. Try Each option, expect it to behave as it previously did. 

## Screenshots 

<img width="360" src="https://user-images.githubusercontent.com/304044/95422027-33b8a900-0947-11eb-8445-661784812595.gif" />

## NOTES

I tried another approach by creating a separate Gutenberg Editor just for the preview functionality ([this is the approach on iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/14871/files#diff-f6c17c64b35f459474509b0698039f10)). This proved quite complex and the amount of code that needed to be duplicated (and maintained) made it a worst approach in my view.
Thus the approach followed in this PR is passing a `isPreview` flag and making the less changes possible to achieve the desired result.

PR submission checklist:
